pkg/lib/authn/identity/loginid/checker.go

```diff
@@ -15,10 +17,17 @@ type Checker struct {
 	TypeCheckerFactory *TypeCheckerFactory
 }
 
-func (c *Checker) ValidateOne(loginID identity.LoginIDSpec, options CheckerOptions) error {
+func (c *Checker) ValidateOne(loginID identity.LoginIDSpec, options CheckerOptions) (identity.LoginIDSpec, error) {
```

Comment:
Having side effects (changing spec.Value) is not good in a function named ValidateOne. Validation normally should be pure. I have the following suggestions

1. Forbid constructing LoginIDSpec directly. Must use NewLoginIDSpec. NewLoginIDSpec calls strings.TrimSpace.
2. Changing the type of LoginIDSpec.Value to UserInputString. This is a safer option, as the compiler will complain on every existing call sites. Existing places have to be modified to `spec.Value.TrimSpace()` to get a safe string to use.

```
type UserInputString struct {
  UnsafeString string
}

func (s UserInputString) TrimSpace() string {
  return strings.TrimSpace(s.UnsafeString)
}

type LoginIDSpec struct {
  Key string
  Type model.LoginIDKey
  Value UserInputString 
}
```

Finally, I think UserInputString can be placed in `pkg/util/userinpututil/string.go`